server.go

```diff
@@ -1494,95 +1471,77 @@ type setDefaultRetentionPolicyCommand struct {
-		}
+func newCreateMeasurementsIfNotExistsCommand(database string) *createMeasurementsIfNotExistsCommand {
+	return &createMeasurementsIfNotExistsCommand{
+		Database:     database,
+		Measurements: make(map[string]createMeasurementSubcommand),
 	}
+}
 
-	// Update Metastore only if the Measurement's fields were actually changed.
-	if len(mm.Fields) > nCurrFields {
-		if err := s.meta.mustUpdate(func(tx *metatx) error {
-			if err := tx.saveMeasurement(db.name, mm); err != nil {
-				return fmt.Errorf("save measurement: %s", err)
-			}
-			return tx.saveDatabase(db)
-		}); err != nil {
-			return err
+// addMeasurementIfNotExists adds the Measurement to the command, but only if not already present
+// in the command.
+func (c *createMeasurementsIfNotExistsCommand) addMeasurementIfNotExists(name string) error {
+	_, ok := c.Measurements[name]
+	if !ok {
+		c.Measurements[name] = createMeasurementSubcommand{
+			Name:   name,
+			Tags:   make(map[string]map[string]string),
+			Fields: make(map[string]influxql.DataType),
 		}
 	}
-
 	return nil
 }
 
-func (s *Server) applyCreateSeriesIfNotExists(m *messaging.Message) error {
-	var c createSeriesIfNotExistsCommand
-	mustUnmarshalJSON(m.Data, &c)
-
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	// Validate command.
-	db := s.databases[c.Database]
-	if db == nil {
-		return ErrDatabaseNotFound
+// addSeriesIfNotExists adds the Series, identified by Measurement name and tag set, to
+// the command, but only if not already present in the command.
+func (c *createMeasurementsIfNotExistsCommand) addSeriesIfNotExists(measurement string, tags map[string]string) error {
+	_, ok := c.Measurements[measurement]
+	if !ok {
+		return ErrMeasurementNotFound
 	}
 
-	if _, series := db.MeasurementAndSeries(c.Name, c.Tags); series != nil {
+	tagset := string(marshalTags(tags))
+	_, ok = c.Measurements[measurement].Tags[tagset]
+	if ok {
+		// Series already present in in subcommand, nothing to do.
 		return nil
 	}
-
-	// save to the metastore and add it to the in memory index
-	var series *Series
-	if err := s.meta.mustUpdate(func(tx *metatx) error {
-		var err error
-		series, err = tx.createSeries(db.name, c.Name, c.Tags)
-		return err
-	}); err != nil {
-		return err
-	}
-
-	db.addSeriesToIndex(c.Name, series)
+	// Tag-set needs to added to subcommand.
+	c.Measurements[measurement].Tags[tagset] = tags
 
 	return nil
 }
 
-type createSeriesIfNotExistsCommand struct {
-	Database string            `json:"database"`
-	Name     string            `json:"name"`
-	Tags     map[string]string `json:"tags"`
+// addFieldIfNotExists adds the field to the command for the Measurement, but only if it is not already
+// present. It will return an error if the field is present in the command, but is of a different type.
+func (c *createMeasurementsIfNotExistsCommand) addFieldIfNotExists(measurement, name string, typ influxql.DataType) error {
+	_, ok := c.Measurements[measurement]
+	if !ok {
+		return ErrMeasurementNotFound
+	}
```

Comment:
Why not keep the reference to `c.Measurements[measurement]` so you can use it below?